<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Felipe Melo">
<meta name="dcterms.date" content="2025-04-09">

<title>Week 2 - Data Wrangling &amp; Research Questions</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/quarto-contrib/fontawesome6-1.2.0/all.min.css" rel="stylesheet">
<link href="../../site_libs/quarto-contrib/fontawesome6-1.2.0/latex-fontsize.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
 
  <style>
  .slide-deck {
      border: 3px solid #dee2e6;
      width: 100%;
  }
  </style>
  


</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../../img/ntu_logo.png" alt="Quarto logo." class="navbar-logo">
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../syllabus.qmd"> 
<span class="menu-text">Syllabus</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../schedule.html"> 
<span class="menu-text">Schedule</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-get-started" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Get Started</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-get-started">    
        <li>
    <a class="dropdown-item" href="../../get-started/1-install.html"><i class="bi bi-emoji-sunglasses-fill" role="img">
</i> 
 <span class="dropdown-text">1 - Install R and Rstudio</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../get-started/2-basic-intro.html"><i class="bi bi-tools" role="img">
</i> 
 <span class="dropdown-text">2 - Basic Introduction to R and Rstudio</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../get-started/3-syntax.html"><i class="bi bi-tools" role="img">
</i> 
 <span class="dropdown-text">3 - Syntax &amp; Data Structures</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../get-started/4-functions.html"><i class="bi bi-tools" role="img">
</i> 
 <span class="dropdown-text">4 - Functions and Arguments</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../get-started/5-data.html"><i class="bi bi-tools" role="img">
</i> 
 <span class="dropdown-text">5 - Reading data in R</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item">
    <a class="nav-link" href="../../sessions/index.html"> 
<span class="menu-text">Content</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../bibliography/index.qmd"> 
<span class="menu-text">Bibliography</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../resources/index.qmd"> 
<span class="menu-text">Resources</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../assessment.html"> 
<span class="menu-text">Assessment</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../docs/blog/index.qmd"> 
<span class="menu-text">Blog</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-help" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Help</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-help">    
        <li>
    <a class="dropdown-item" href="https://github.com/quarto-dev/quarto-cli/issues"><i class="bi bi-bug" role="img">
</i> 
 <span class="dropdown-text">Report a Bug</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="https://github.com/quarto-dev/quarto-cli/discussions"><i class="bi bi-chat-right-text" role="img">
</i> 
 <span class="dropdown-text">Ask a Question</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../docs/faq/index.qmd"><i class="bi bi-question-circle" role="img">
</i> 
 <span class="dropdown-text">FAQ</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools tools-wide">
    <a href="https://twitter.com/quarto_pub" title="Quarto Twitter" class="quarto-navigation-tool px-1" aria-label="Quarto Twitter"><i class="bi bi-twitter"></i></a>
    <a href="https://github.com/quarto-dev/quarto-cli" title="Quarto GitHub" class="quarto-navigation-tool px-1" aria-label="Quarto GitHub"><i class="bi bi-github"></i></a>
    <a href="https://quarto.org/docs/blog/index.xml" title="Quarto Blog RSS" class="quarto-navigation-tool px-1" aria-label="Quarto Blog RSS"><i class="bi bi-rss"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#preparation" id="toc-preparation" class="nav-link active" data-scroll-target="#preparation">Preparation</a>
  <ul class="collapse">
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link" data-scroll-target="#learning-objectives">Learning objectives</a></li>
  <li><a href="#prerequisites" id="toc-prerequisites" class="nav-link" data-scroll-target="#prerequisites">Prerequisites</a></li>
  <li><a href="#readings" id="toc-readings" class="nav-link" data-scroll-target="#readings">Readings</a></li>
  </ul></li>
  <li><a href="#lesson" id="toc-lesson" class="nav-link" data-scroll-target="#lesson">Lesson</a></li>
  <li><a href="#exercise" id="toc-exercise" class="nav-link" data-scroll-target="#exercise">Exercise</a>
  <ul class="collapse">
  <li><a href="#learning-objectives-1" id="toc-learning-objectives-1" class="nav-link" data-scroll-target="#learning-objectives-1">Learning Objectives</a></li>
  </ul></li>
  <li><a href="#data-wrangling-with-tidyverse" id="toc-data-wrangling-with-tidyverse" class="nav-link" data-scroll-target="#data-wrangling-with-tidyverse">Data Wrangling with Tidyverse</a>
  <ul class="collapse">
  <li><a href="#tidyverse-basics" id="toc-tidyverse-basics" class="nav-link" data-scroll-target="#tidyverse-basics">Tidyverse basics</a>
  <ul class="collapse">
  <li><a href="#pipes" id="toc-pipes" class="nav-link" data-scroll-target="#pipes">Pipes</a></li>
  <li><a href="#tibbles" id="toc-tibbles" class="nav-link" data-scroll-target="#tibbles">Tibbles</a></li>
  </ul></li>
  <li><a href="#experimental-data" id="toc-experimental-data" class="nav-link" data-scroll-target="#experimental-data">Experimental data</a></li>
  <li><a href="#analysis-goal-and-workflow" id="toc-analysis-goal-and-workflow" class="nav-link" data-scroll-target="#analysis-goal-and-workflow">Analysis goal and workflow</a></li>
  <li><a href="#tidyverse-tools" id="toc-tidyverse-tools" class="nav-link" data-scroll-target="#tidyverse-tools">Tidyverse tools</a></li>
  <li><a href="#read-in-the-functional-analysis-results" id="toc-read-in-the-functional-analysis-results" class="nav-link" data-scroll-target="#read-in-the-functional-analysis-results">1. Read in the functional analysis results</a></li>
  <li><a href="#extract-only-the-go-biological-processes-bp-of-interest" id="toc-extract-only-the-go-biological-processes-bp-of-interest" class="nav-link" data-scroll-target="#extract-only-the-go-biological-processes-bp-of-interest">2. Extract only the GO biological processes (BP) of interest</a></li>
  <li><a href="#select-only-the-columns-needed-for-visualization" id="toc-select-only-the-columns-needed-for-visualization" class="nav-link" data-scroll-target="#select-only-the-columns-needed-for-visualization">3. Select only the columns needed for visualization</a></li>
  <li><a href="#order-go-processes-by-significance-adjusted-p-values" id="toc-order-go-processes-by-significance-adjusted-p-values" class="nav-link" data-scroll-target="#order-go-processes-by-significance-adjusted-p-values">4. Order GO processes by significance (adjusted p-values)</a></li>
  <li><a href="#rename-columns-to-be-more-intuitive" id="toc-rename-columns-to-be-more-intuitive" class="nav-link" data-scroll-target="#rename-columns-to-be-more-intuitive">5. Rename columns to be more intuitive</a></li>
  <li><a href="#create-additional-metrics-for-plotting-e.g.-gene-ratios" id="toc-create-additional-metrics-for-plotting-e.g.-gene-ratios" class="nav-link" data-scroll-target="#create-additional-metrics-for-plotting-e.g.-gene-ratios">6. Create additional metrics for plotting (e.g.&nbsp;gene ratios)</a></li>
  <li><a href="#next-steps" id="toc-next-steps" class="nav-link" data-scroll-target="#next-steps">Next steps</a>
  <ul class="collapse">
  <li><a href="#additional-resources" id="toc-additional-resources" class="nav-link" data-scroll-target="#additional-resources">Additional resources</a></li>
  </ul></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content column-page-left" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Week 2 - Data Wrangling &amp; Research Questions</h1>
<p class="subtitle lead">RMDA ARES40011</p>
</div>



<div class="quarto-title-meta column-page-left">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Felipe Melo </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">April 9, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<p><img src="../../img/dplyr_wrangling.png" class="img-fluid" width="500"></p>
<p><a href="https://twitter.com/allison_horst" style="text-align: center">Artwork by @allison_horst</a></p>
<section id="preparation" class="level1">
<h1>Preparation</h1>
<section id="learning-objectives" class="level2">
<h2 class="anchored" data-anchor-id="learning-objectives">Learning objectives</h2>
<table class="caption-top table">
<colgroup>
<col style="width: 50%">
<col style="width: 50%">
</colgroup>
<thead>
<tr class="header">
<th>Research Methods</th>
<th>Data Analyses</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Reflect on how to make good research questions</td>
<td>Manipulate datasets as needed</td>
</tr>
<tr class="even">
<td></td>
<td>Use <strong>Tidyverse</strong> R package</td>
</tr>
<tr class="odd">
<td></td>
<td>Produce simple tables and summaries</td>
</tr>
</tbody>
</table>
</section>
<section id="prerequisites" class="level2">
<h2 class="anchored" data-anchor-id="prerequisites">Prerequisites</h2>
<p>Before you begin, ensure you have the following installed:</p>
<ul>
<li><p><strong>R</strong>: You can download the latest version from the official R website. Instructions for different operating systems (Windows, Mac OS, UNIX/Linux) are available [8]. <strong>R must be installed before installing RStudio</strong> [9].</p></li>
<li><p><strong>RStudio</strong>: Download and install RStudio Desktop from <a href="https://www.rstudio.com/" class="uri">https://www.rstudio.com/</a> [10]. To open RStudio, locate it in your applications and launch it [11].</p></li>
<li><p><strong>tidyverse package</strong>: Install the <strong>tidyverse</strong> package within R. Open RStudio and in the <strong>Console</strong> pane, type and run the following command [12]:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"tidyverse"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
</ul>
<p>This command needs to be run <strong>only the first time</strong> you want to use the <strong>tidyverse</strong>. You can also install packages by navigating to <strong>Tools</strong> -&gt; <strong>Install Packages</strong> in RStudio.</p>
</section>
<section id="readings" class="level2">
<h2 class="anchored" data-anchor-id="readings">Readings</h2>
<ul>
<li><i class="fa-solid fa-book" aria-label="book"></i> Check chapters 1 to 3 of the e-book <a href="https://jhudatascience.org/tidyversecourse/intro.html">Tidyverse Skills for Data Science</a></li>
<li><i class="fa-solid fa-tools" aria-label="tools"></i> <a href="https://rstudio.github.io/cheatsheets/html/tidyr.html?_gl=1*17wj54u*_ga*NzI5NTAxMDIzLjE3NDI0Njc1OTk.*_ga_2C0WZ1JHG0*MTc0NDAyNDc5My4zLjEuMTc0NDAyNDgzNC4wLjAuMA..">Data tidying with tidyr :: Cheatsheet</a></li>
<li><i class="fa-solid fa-tools" aria-label="tools"></i> <a href="https://rstudio.github.io/cheatsheets/html/data-transformation.html?_gl=1*ny4r9m*_ga*NzI5NTAxMDIzLjE3NDI0Njc1OTk.*_ga_2C0WZ1JHG0*MTc0NDAyNDc5My4zLjEuMTc0NDAyNDgzNC4wLjAuMA..">Data transformation with dplyr :: Cheatsheet</a></li>
</ul>
</section>
</section>
<section id="lesson" class="level1">
<h1>Lesson</h1>
    <p><a href="https://fplmelo.github.io/rmda_ntu/slides/02_data_wrangling.html#/title-slide" target="_blank">View slides in full screen</a></p>
       <div class="ratio ratio-16x9">
      <iframe class="slide-deck" src="https://fplmelo.github.io/rmda_ntu/slides/02_data_wrangling.html#/title-slide" height="500px"></iframe>
    </div>
  
    <p><a href="https://fplmelo.github.io/rmda_ntu/slides/2.2_research_question.html#/title-slide" target="_blank">View slides in full screen</a></p>
       <div class="ratio ratio-16x9">
      <iframe class="slide-deck" src="https://fplmelo.github.io/rmda_ntu/slides/2.2_research_question.html#/title-slide" height="500px"></iframe>
    </div>
  
</section>
<section id="exercise" class="level1">
<h1>Exercise</h1>
<p>Approximate time: 60 min</p>
<section id="learning-objectives-1" class="level2">
<h2 class="anchored" data-anchor-id="learning-objectives-1">Learning Objectives</h2>
<ul>
<li>Demonstrate how to subset, merge, and create new datasets from existing data structures in R.</li>
<li>Perform basic data wrangling with functions in the Tidyverse package.</li>
</ul>
</section>
</section>
<section id="data-wrangling-with-tidyverse" class="level1">
<h1>Data Wrangling with Tidyverse</h1>
<p>The <a href="https://www.tidyverse.org/packages/">Tidyverse suite of integrated packages</a> are designed to work together to make common data science operations more user friendly. The packages have functions for data wrangling, tidying, reading/writing, parsing, and visualizing, among others. There is a freely available book, <a href="http://r4ds.had.co.nz/index.html">R for Data Science</a>, with detailed descriptions and practical examples of the tools available and how they work together. We will explore the basic syntax for working with these packages, as well as, specific functions for data wrangling with the ‘dplyr’ package and data visualization with the ‘ggplot2’ package.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../../img/tidyverse_website.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></p>
</figure>
</div>
<section id="tidyverse-basics" class="level2">
<h2 class="anchored" data-anchor-id="tidyverse-basics">Tidyverse basics</h2>
<p>The Tidyverse suite of packages introduces users to a set of data structures, functions and operators to make working with data more intuitive, but is slightly different from the way we do things in base R. <strong>Two important new concepts we will focus on are pipes and tibbles</strong>.</p>
<p>Before we get started with pipes or tibbles, let’s load the library:</p>
<pre><code>library(tidyverse)</code></pre>
<section id="pipes" class="level3">
<h3 class="anchored" data-anchor-id="pipes">Pipes</h3>
<p>Stringing together commands in R can be quite daunting. Also, trying to understand code that has many nested functions can be confusing.</p>
<p>To make R code more human readable, the Tidyverse tools use the pipe, <code>%&gt;%</code>, which was acquired from the <code>magrittr</code> package and is now part of the <code>dplyr</code> package that is installed automatically with Tidyverse. <strong>The pipe allows the output of a previous command to be used as input to another command instead of using nested functions.</strong></p>
<blockquote class="blockquote">
<p><strong>NOTE:</strong> Shortcut to write the pipe is <kbd>shift</kbd> + <kbd>command/ctrl</kbd> + <kbd>M</kbd></p>
</blockquote>
<p>An example of using the pipe to run multiple commands:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="do">## A single command</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sqrt</span>(<span class="dv">83</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Base R method of running more than one command</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">sqrt</span>(<span class="dv">83</span>), <span class="at">digits =</span> <span class="dv">2</span>)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Running more than one command with piping</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="fu">sqrt</span>(<span class="dv">83</span>) <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="at">digits =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>The pipe represents a much easier way of writing and deciphering R code, and so we will be taking advantage of it, when possible, as we work through the remaining lesson.</p>
<hr>
<p><strong>Exercises</strong></p>
<ol type="1">
<li><p>Create a vector of random numbers using the code below:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>random_numbers <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">81</span>, <span class="dv">90</span>, <span class="dv">65</span>, <span class="dv">43</span>, <span class="dv">71</span>, <span class="dv">29</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div></li>
<li><p>Use the pipe (<code>%&gt;%</code>) to perform two steps in a single line:</p>
<ol type="1">
<li>Take the mean of <code>random_numbers</code> using the <code>mean()</code> function.</li>
<li>Round the output to three digits using the <code>round()</code> function.</li>
</ol></li>
</ol>
<hr>
</section>
<section id="tibbles" class="level3">
<h3 class="anchored" data-anchor-id="tibbles">Tibbles</h3>
<p>A core component of the <a href="http://tidyverse.org/">tidyverse</a> is the <a href="http://tibble.tidyverse.org/">tibble</a>. <strong>Tibbles are a modern rework of the standard <code>data.frame</code>, with some internal improvements</strong> to make code more reliable. They are data frames, but do not follow all of the same rules. For example, tibbles can have numbers/symbols for column names, which is not normally allowed in base R.</p>
<p><strong>Important: <a href="http://tidyverse.org/">tidyverse</a> is very opininated about row names</strong>. These packages insist that all column data (e.g.&nbsp;<code>data.frame</code>) be treated equally, and that special designation of a column as <code>rownames</code> should be deprecated. <a href="http://tibble.tidyverse.org/">Tibble</a> provides simple utility functions to handle rownames: <code>rownames_to_column()</code> and <code>column_to_rownames()</code>.</p>
<p>Tibbles can be created directly using the <code>tibble()</code> function or data frames can be converted into tibbles using <code>as_tibble(name_of_df)</code>.</p>
<blockquote class="blockquote">
<p><strong>NOTE:</strong> The function <code>as_tibble()</code> will ignore row names, so if a column representing the row names is needed, then the function <code>rownames_to_column(name_of_df)</code> should be run prior to turning the data.frame into a tibble. Also, <code>as_tibble()</code> will not coerce character vectors to factors by default.</p>
</blockquote>
</section>
</section>
<section id="experimental-data" class="level2">
<h2 class="anchored" data-anchor-id="experimental-data">Experimental data</h2>
<p>We’re going to explore the Tidyverse suite of tools to wrangle our data to prepare it for visualization. You should have downloaded the file called <code>gprofiler_results_Mov10oe.tsv</code> into your R project’s <code>data</code> folder earlier.</p>
<blockquote class="blockquote">
<p>If you do not have the <code>gprofiler_results_Mov10oe.tsv</code> file in your <code>data</code> folder, you can right click and download it into the <code>data</code> folder <a href="https://github.com/hbctraining/Training-modules/blob/master/Tidyverse_ggplot2/data/gprofiler_results_Mov10oe.tsv?raw=true">using this link</a>.</p>
</blockquote>
<p><strong>The dataset:</strong></p>
<ul>
<li>Represents the <strong>functional analysis results</strong>, including the biological processes, functions, pathways, or conditions that are over-represented in a given list of genes.</li>
<li>Our gene list was generated by <strong>differential gene expression analysis</strong> and the genes represent differences between <strong>control mice</strong> and <strong>mice over-expressing a gene involved in RNA splicing</strong>.</li>
</ul>
<p>The functional analysis that we will focus on involves <strong>gene ontology (GO) terms</strong>, which:</p>
<ul>
<li>describe the roles of genes and gene products</li>
<li>organized into three controlled vocabularies/ontologies (domains):
<ul>
<li>biological processes (BP)</li>
<li>cellular components (CC)</li>
<li>molecular functions (MF)</li>
</ul></li>
</ul>
<p><img src="../../img/mov10_FA.png" width="1200"></p>
</section>
<section id="analysis-goal-and-workflow" class="level2">
<h2 class="anchored" data-anchor-id="analysis-goal-and-workflow">Analysis goal and workflow</h2>
<p><strong>Goal:</strong> <em>Visually compare the most significant biological processes (BP) based on the number of associated differentially expressed genes (gene ratios) and significance values by creating the following plot:</em></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../../img/dotplot6.png" class="img-fluid figure-img"></p>
<figcaption>dotplot6</figcaption>
</figure>
</div>
<p>To wrangle our data in preparation for the plotting, we are going to use the Tidyverse suite of tools to wrangle and visualize our data through several steps:</p>
<ol type="1">
<li>Read in the functional analysis results</li>
<li>Extract only the GO biological processes (BP) of interest</li>
<li>Select only the columns needed for visualization</li>
<li>Order by significance (p-adjusted values)</li>
<li>Rename columns to be more intuitive</li>
<li>Create additional metrics for plotting (e.g.&nbsp;gene ratios)</li>
<li>Plot results</li>
</ol>
</section>
<section id="tidyverse-tools" class="level2">
<h2 class="anchored" data-anchor-id="tidyverse-tools">Tidyverse tools</h2>
<p>While all of the tools in the Tidyverse suite are deserving of being explored in more depth, we are going to investigate more deeply the reading (<code>readr</code>), wrangling (<code>dplyr</code>), and plotting (<code>ggplot2</code>) tools.</p>
</section>
<section id="read-in-the-functional-analysis-results" class="level2">
<h2 class="anchored" data-anchor-id="read-in-the-functional-analysis-results">1. Read in the functional analysis results</h2>
<p>While the base R packages have perfectly fine methods for reading in data, the <code>readr</code> and <code>readxl</code> Tidyverse packages offer additional methods for reading in data. Let’s read in our tab-delimited functional analysis results using <code>read_delim()</code>:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read in the functional analysis results</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>functional_GO_results <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(<span class="at">file =</span> <span class="st">"data/gprofiler_results_Mov10oe.tsv"</span>, <span class="at">delim =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span> )</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look at the results</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>functional_GO_results</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<details>
<summary>
<i>Click here to see how to do this in base R</i>
</summary>
<p><br></p>
<pre><code># Read in the functional analysis results<br>
functional_GO_results &lt;- read.delim(file = "data/gprofiler_results_Mov10oe.tsv", sep = "\t" )</code><br>
<code># Take a look at the results
functional_GO_results</code></pre>
<p><br></p>
</details>
<p>Notice that the results were automatically read in as a tibble and the output gives the number of rows, columns and the data type for each of the columns.</p>
<blockquote class="blockquote">
<p><strong>NOTE</strong>: A large number of tidyverse functions will work with both tibbles and dataframes, and the data structure of the output will be identical to the input. However, there are some functions that will return a tibble (without row names), whether or not a tibble or dataframe is provided.</p>
</blockquote>
</section>
<section id="extract-only-the-go-biological-processes-bp-of-interest" class="level2">
<h2 class="anchored" data-anchor-id="extract-only-the-go-biological-processes-bp-of-interest">2. Extract only the GO biological processes (BP) of interest</h2>
<p>Now that we have our data, we will need to wrangle it into a format ready for plotting. For all of our data wrangling steps we will be using tools from the <a href="http://dplyr.tidyverse.org/">dplyr</a> package, which is a swiss-army knife for data wrangling of data frames.</p>
<p>To extract the biological processes of interest, we only want those rows where the <code>domain</code> is equal to <code>BP</code>, which we can do using the <code>filter()</code> function.</p>
<p>To filter rows of a data frame/tibble based on values in different columns, we give a logical expression as input to the <code>filter()</code> function to return those rows for which the expression is TRUE.</p>
<p>Now let’s return only those rows that have a <code>domain</code> of <code>BP</code>:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Return only GO biological processes</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> functional_GO_results <span class="sc">%&gt;%</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(domain <span class="sc">==</span> <span class="st">"BP"</span>)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(bp_oe)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<details>
<summary>
<i>Click here to see how to do this in base R</i>
</summary>
<p><br></p>
<pre><code># Return only GO biological processes
idx &lt;- functional_GO_results$domain == "BP"
bp_oe2 &lt;- functional_GO_results[idx,]</code><br>
<code>View(bp_oe)</code></pre>
<p><br></p>
</details>
<p>Now we have returned only those rows with a <code>domain</code> of <code>BP</code>. <strong>How have the dimensions of our results changed?</strong></p>
<hr>
<p><strong>Exercise:</strong></p>
<p>We would like to perform an additional round of filtering to only keep the most specific GO terms.</p>
<ol type="1">
<li>For <code>bp_oe</code>, use the <code>filter()</code> function to only keep those rows where the <code>relative.depth</code> is greater than 4.</li>
<li>Save output to overwrite our <code>bp_oe</code> variable.</li>
</ol>
<hr>
</section>
<section id="select-only-the-columns-needed-for-visualization" class="level2">
<h2 class="anchored" data-anchor-id="select-only-the-columns-needed-for-visualization">3. Select only the columns needed for visualization</h2>
<p>For visualization purposes, we are only interested in the columns related to the GO terms, the significance of the terms, and information about the number of genes associated with the terms.</p>
<p>To extract columns from a data frame/tibble we can use the <code>select()</code> function. In contrast to base R, we do not need to put the column names in quotes for selection.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Selecting columns to keep</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> bp_oe <span class="sc">%&gt;%</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(term.id, term.name, p.value, query.size, term.size, overlap.size, intersection)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="fu">View</span>(bp_oe)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<details>
<summary>
<i>Click here to see how to do this in base R</i>
</summary>
<p><br></p>
<pre><code># Selecting columns to keep
bp_oe &lt;- bp_oe[, c("term.id", "term.name", "p.value", "query.size", "term.size", "overlap.size", "intersection")]</code><br>
<code>View(bp_oe)</code></pre>
<p><br></p>
</details>
<p>The <code>select()</code> function also allows for negative selection. So we could have alternately removed columns with negative selection. Note that we need to put the column names inside of the combine (<code>c()</code>) function with a <code>-</code> preceding it for this functionality.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Selecting columns to remove</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> bp_oe <span class="sc">%&gt;%</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">select</span>(<span class="sc">-</span><span class="fu">c</span>(query.number, significant, recall, precision, subgraph.number, relative.depth, domain))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<details>
<summary>
<i>Click here to see how to do this in base R</i>
</summary>
<p><br></p>
<pre><code># DO NOT RUN</code>
<code># Selecting columns to remove
idx &lt;- !(colnames(bp_oe) %in% c("query.number", "significant", "recall", "precision", "subgraph.number", "relative.depth", "domain"))
bp_oe &lt;- bp_oe[, idx]</code></pre>
<p><br></p>
</details>
<p><img src="../../img/bp_oe_selection.png" width="1200"></p>
</section>
<section id="order-go-processes-by-significance-adjusted-p-values" class="level2">
<h2 class="anchored" data-anchor-id="order-go-processes-by-significance-adjusted-p-values">4. Order GO processes by significance (adjusted p-values)</h2>
<p>Now that we have only the rows and columns of interest, let’s arrange these by significance, which is denoted by the adjusted p-value.</p>
<p>Let’s sort the rows by adjusted p-value with the <code>arrange()</code> function.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Order by adjusted p-value ascending</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> bp_oe <span class="sc">%&gt;%</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(p.value)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<details>
<summary>
<i>Click here to see how to do this in base R</i>
</summary>
<p><br></p>
<pre><code># Order by adjusted p-value ascending
idx &lt;- order(bp_oe$p.value)
bp_oe &lt;- bp_oe[idx,]</code></pre>
<p><br></p>
</details>
<blockquote class="blockquote">
<p><strong>NOTE1:</strong> If you wanted to arrange in descending order, then you could have run the following instead:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># DO NOT RUN</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Order by adjusted p-value descending</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> bp_oe <span class="sc">%&gt;%</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="fu">desc</span>(p.value))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<details>
<summary>
<i>Click here to see how to do this in base R</i>
</summary>
<p><br></p>
<pre><code># DO NOT RUN</code>
<code># Order by adjusted p-value descending
idx &lt;- order(bp_oe$p.value, decreasing = TRUE)
bp_oe &lt;- bp_oe[idx,]</code></pre>
</details>
</blockquote>
<blockquote class="blockquote">
<p><strong>NOTE2:</strong> Ordering variables in <code>ggplot2</code> is a bit different. <a href="https://www.r-graph-gallery.com/267-reorder-a-variable-in-ggplot2.html">This post</a> introduces a few ways of ordering variables in a plot.</p>
</blockquote>
</section>
<section id="rename-columns-to-be-more-intuitive" class="level2">
<h2 class="anchored" data-anchor-id="rename-columns-to-be-more-intuitive">5. Rename columns to be more intuitive</h2>
<p>While not necessary for our visualization, renaming columns more intuitively can help with our understanding of the data using the <code>rename()</code> function. The syntax is <code>new_name</code> = <code>old_name</code>.</p>
<p>Let’s rename the <code>term.id</code> and <code>term.name</code> columns.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Provide better names for columns</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> bp_oe <span class="sc">%&gt;%</span> </span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">rename</span>(<span class="at">GO_id =</span> term.id, </span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">GO_term =</span> term.name)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<details>
<summary>
<i>Click here to see how to do this in base R</i>
</summary>
<p><br></p>
<pre><code># Provide better names for columns
colnames(bp_oe)[colnames(bp_oe) == "term.id"] &lt;- "GO_id"
colnames(bp_oe)[colnames(bp_oe) == "term.name"] &lt;- "GO_term"</code></pre>
<p><br></p>
</details>
<blockquote class="blockquote">
<p><strong>NOTE:</strong> In the case of two packages with identical function names, you can use <code>::</code> with the package name before and the function name after (e.g <code>stats::filter()</code>) to ensure that the correct function is implemented. The <code>::</code> can also be used to bring in a function from a library without loading it first.</p>
<p>In the example above, we wanted to use the <code>rename()</code> function specifically from the <code>dplyr</code> package, and not any of the other packages (or base R) which may have the <code>rename()</code> function.</p>
</blockquote>
<hr>
<p><strong>Exercise</strong></p>
<p>Rename the <code>intersection</code> column to <code>genes</code> to reflect the fact that these are the DE genes associated with the GO process.</p>
<hr>
</section>
<section id="create-additional-metrics-for-plotting-e.g.-gene-ratios" class="level2">
<h2 class="anchored" data-anchor-id="create-additional-metrics-for-plotting-e.g.-gene-ratios">6. Create additional metrics for plotting (e.g.&nbsp;gene ratios)</h2>
<p>Finally, before we plot our data, we need to create a couple of additional metrics. The <code>mutate()</code> function enables you to create a new column from an existing column.</p>
<p>Let’s generate gene ratios to reflect the number of DE genes associated with each GO process relative to the total number of DE genes.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create gene ratio column based on other columns in dataset</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>bp_oe <span class="ot">&lt;-</span> bp_oe <span class="sc">%&gt;%</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">gene_ratio =</span> overlap.size <span class="sc">/</span> query.size)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<details>
<summary>
<i>Click here to see how to do this in base R</i>
</summary>
<p><br></p>
<pre><code># Create gene ratio column based on other columns in dataset
bp_oe &lt;- cbind(bp_oe, gene_ratio = bp_oe$overlap.size / bp_oe$query.size)</code></pre>
<p><br></p>
</details>
<hr>
<p><strong>Exercise</strong></p>
<p>Create a column in <code>bp_oe</code> called <code>term_percent</code> to determine the percent of DE genes associated with the GO term relative to the total number of genes associated with the GO term (<code>overlap.size</code> / <code>term.size</code>)</p>
<hr>
<p>Our final data for plotting should look like the table below:</p>
<p><img src="../../img/bp_oe_ready_to_plot.png" width="1200"></p>
</section>
<section id="next-steps" class="level2">
<h2 class="anchored" data-anchor-id="next-steps">Next steps</h2>
<p>Now that we have our results ready for plotting, we can use the <a href="https://ggplot2.tidyverse.org">ggplot2</a> package to plot our results. If you are interested, you can follow <a href="https://hbctraining.github.io/Training-modules/Tidyverse_ggplot2/lessons/03_ggplot2.html">this lesson</a> and dive into how to use <code>ggplot2</code> to create the plots with this dataset.</p>
<section id="additional-resources" class="level3">
<h3 class="anchored" data-anchor-id="additional-resources">Additional resources</h3>
<ul>
<li><a href="http://r4ds.had.co.nz">R for Data Science</a></li>
<li><a href="http://varianceexplained.org/r/teach-tidyverse/">teach the tidyverse</a></li>
<li><a href="http://style.tidyverse.org/">tidy style guide</a></li>
</ul>
<hr>
<p><em>This lesson has been developed by members of the teaching team at the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core (HBC)</a>. These are open access materials distributed under the terms of the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.</em></p>


</section>
</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>Content <i class="fa-brands fa-creative-commons" aria-label="creative-commons"></i> <strong>?var:course.copyright_year</strong> by <a href="?var:instructor.url"><strong>?var:instructor.name_no_title</strong></a> <br> All content licensed under a <i class="fa-brands fa-creative-commons" aria-label="creative-commons"></i> <i class="fa-brands fa-creative-commons-by" aria-label="creative-commons-by"></i> <i class="fa-brands fa-creative-commons-nc" aria-label="creative-commons-nc"></i> <a href="https://creativecommons.org/licenses/by-nc/4.0/">Creative Commons Attribution-NonCommercial 4.0 International license (CC BY-NC 4.0)</a></p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>Made with <i class="fa-brands fa-r-project" aria-label="r-project"></i> and <a href="https://quarto.org/">Quarto</a><br> <a href="?var:course.github">View the source at <i class="fa-brands fa-github" aria-label="github"></i> GitHub</a></p>
</div>
  </div>
</footer>




</body></html>